src/ruleset_builder.cpp

```diff
@@ -202,15 +242,39 @@ ruleset_builder::change_state ruleset_builder::load(parameter::map &root, rulese
         auto new_base_rules =
             parser::v2::parse_rules(rules, info, target_manifest_, rule_data_ids, limits_);
 
-        if (new_base_rules.empty()) {
-            throw ddwaf::parsing_error("no valid rules found");
+        if (!new_base_rules.empty()) {
+            // Upon reaching this stage, we know our base ruleset is valid
+            base_rules_ = std::move(new_base_rules);
+            rule_data_ids_ = std::move(rule_data_ids);
+            state = state | change_state::rules;
         }
+    }
+
+    it = root.find("custom_rules");
+    if (it != root.end()) {
+        // Rule data is currently not supported by custom rules so these will
+        // be discarded after
+        decltype(rule_data_ids_) rule_data_ids;
+
+        auto rules = static_cast<parameter::vector>(it->second);
+        if (!rules.empty()) {
+            auto new_user_rules = parser::v2::parse_rules(
+                rules, info, target_manifest_, rule_data_ids, limits_, rule::source_type::user);
+
+            if (new_user_rules.empty()) {
+                DDWAF_WARN("No valid custom rules found");
+            } else {
+                // Upon reaching this stage, we know our base ruleset is valid
+                user_rules_ = std::move(new_user_rules);
+                state = state | change_state::custom_rules;
```

Comment:
If I'm reading this right, there's kind of an asymmetry in that if we already have rules, these are replaced if we have one invalid rule and one valid rule, but not if we have only one invalid rule. IMO, the old rules should be discarded in both cases.

Reply:
Thinking about it, I guess that's fair, otherwise it might be difficult to understand the state of the WAF.

Reply:
I've fixed this asymmetry for all top-level keys.